python/rmm/_lib/device_buffer.pyx

```diff
@@ -139,3 +152,42 @@ cdef class DeviceBuffer:
 
     cdef void* c_data(self):
         return self.c_obj.get()[0].data()
+
+
+@cython.boundscheck(False)
+cpdef DeviceBuffer to_device(const unsigned char[::1] b, uintptr_t stream=0):
+    if b is None:
+        raise TypeError(
+            "Argument 'b' has incorrect type"
+            " (expected bytes, got NoneType)"
+        )
+
+    cdef uintptr_t p = <uintptr_t>&b[0]
+    cdef size_t s = len(b)
+    return DeviceBuffer(ptr=p, size=s, stream=stream)
+
+
+@cython.boundscheck(False)
+cpdef void copy_to_host(uintptr_t db,
+                        unsigned char[::1] hb,
+                        uintptr_t stream=0) nogil except *:
+    if hb is None:
+        with gil:
+            raise TypeError(
+                "Argument `hb` has incorrect type"
+                " (expected bytes-like, got NoneType)"
+            )
+
+    cdef cudaError_t err
+
+    err = cudaMemcpyAsync(<void*>&hb[0], <const void*>db, len(hb),
+                          cudaMemcpyDeviceToHost, <cudaStream_t>stream)
+    if err != cudaSuccess:
+        with gil:
+            raise RuntimeError(f"Memcpy failed with error: {err}")
+
+    if stream == 0:
+        err = cudaStreamSynchronize(<cudaStream_t>stream)
```

Comment:
After discussion with Keith offline, we decided to synchronize explicitly on the default stream, but not on any other streams. Have added this code above.

Reply:
Why? If the stream is 0, `cudaMemcpyAsync` will act like a synchronous `cudaMemcpy`.

Reply:
This is what Keith and I were unclear about. Could you please point us to the docs where it says this?

Reply:
I glossed over some details in an attempt at brevity, but in all honesty, default stream behavior is a mess. 

The gory detail is that when we're talking about the synchronization behavior CUDA runtime APIs we have to be clear about if a function is synchronous with respect to the _host_ or _device_, i.e., if a function blocks forward progress on the host or device.

So `cudaMemcpyAsync` in the default (zero) stream _is_ different from `cudaMemcpy` in that `cudaMemcpyAsync` is always asynchronous wrt to the host thread. I.e., `cudaMemcpyAsync(...,0)` returns control to the host thread immediately, even if the copy is not yet complete. 

Example:
```c++
cudaMemcpy(a_d, a_h, size, cudaMemcpyHostToDevice);
kernel<<<grid, block>>>(a_d);
cudaMemcpy(a_h, a_d, size, cudaMemcpyDeviceToHost);
cpuFunction();
```
In this example `cpuFunction()` cannot execute until the H2D copy, kernel, and D2H copy is complete. This is because `cudaMemcpy` is synchronous wrt to the host.

```c++
cudaMemcpyAsync(a_d, a_h, size, cudaMemcpyHostToDevice, 0);
kernel<<<grid, block>>>(a_d);
cudaMemcpyAsync(a_h, a_d, size, cudaMemcpyDeviceToHost, 0);
cpuFunction();
```

In this example, `cpuFunction()` can overlap with the H2D copy, the kernel, and the D2H copy. This is because `cudaMemcpyAsync` is asynchronous wrt to the host (even on the default stream). 

The problem with `cudaStreamSynchronize(0)` is that stream 0 is special in that it synchronizes _all_ streams(*). I.e., it is effectively the same as a `cudaDeviceSynchronize()`. 

Example:
```c++
kernel1<<<..., stream1>>>(...);
cudaStreamSynchronize(0); 
kernel2<<<..., stream2>>>(...);
```

In this example `kernel1` cannot overlap with `kernel2` because of [the synchronization stream 0 implicitly syncs all other streams](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#implicit-synchronization).

Consider instead:
```c++
kernel1<<<..., stream1>>>(...);
cudaStreamSynchronize(stream3); 
kernel2<<<..., stream2>>>(...);
```

Now, `kernel1` and `kernel2` _may_ overlap. 

(*) Unless the application is compiled with [`--default-stream per-thread`](https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/). In which case the default stream is no different from a normal stream.

Reply:
@jrhemstad the fact that `cpuFunction()` can overlap is exactly what we need to guard against. It would be super common for someone to do something like:

```
my_db = rmm.DeviceBuffer(size=100)
some_gpu_function(my_db) # Modifies memory of device buffer
my_hb = my_db.copy_to_host()
some_cpu_function(my_hb) 
```

Where we need the D --> H copy to be guaranteed to be finished before running `some_cpu_function`.

Reply:
It seems very dangerous to have an API that is _sometimes_ synchronous. I'd think it should either be always asynchronous (and require user to synchronize the stream) or always synchronous.

Reply:
I'm not sure I agree, but this may be a difference in language philosophy (see below). 🙂

This matches similar design decisions to that of CuPy and Numba. It was also requested by Peter earlier in the PR. So it sounds like this is an API users expect in Python.

By default, we make sure users are not shooting themselves in the foot, which seems safe. However we leave the door open for the power features, in which case users need to be responsible. This more-or-less aligns with the culture of Python (one of consenting adults 😉)

Reply:
+1 @jakirkham this is the expectation across GPU Python libraries as of now. Whether it is "good" or not remains to be seen. We could eventually expose some type of future based interface but that would be a very non-trivial amount of work, and even more work to get the community to standardize on it.

Reply:
So to summarize, you are saying that if the stream is 0, you are assuming that the user is not stream savvy, and so you synchronize the entire device.  If the stream is non-zero, you assume the user is stream savvy, and you do no synchronization?  This definitely requires documentation.

Reply:
In my opinion, RMM should only be an internal detail of RAPIDS libraries, and not something that RAPIDS users interact with.

Reply:
No objections to additional documentation. What would you propose?

Well we are already using it with things outside of RAPIDS like CuPy ( https://github.com/rapidsai/rmm/pull/198 ) and there are plans to use it with Numba ( https://github.com/rapidsai/rmm/issues/180#issuecomment-578889668 ). Meeting the communities expectations should help make it easier to use with these external libraries.